# Практическая работа №1: Исследование алгоритмов формирования аддитивных цепочек

Выполнил студент гр. 0391 Бысов Федор, вариант 53.

## Цель работы

Формирование представления о аддитивных цепочках, выработать умение составлять и применять алгоритмы для нахождения минимальных аддитивных цепочек для заданного числа, привить навык использования систем компьютерной математики для реализации алгоритмов.

## Основные теоретические положения

Аддитивной цепочкой для натурального числа  называется последовательность натуральных чисел 
, где каждый элемент последовательности равен сумме каких-либо двух предыдущих:

## Основные теоретические положения

$\quad$ Аддитивная цепочка - последовательность натуральных чисел от 1 до n
$$ 1 = a_0, a_1, ..., a_r = n ,$$ 
где каждый элемент равен сумме каких-либо двух предыдущих элементов
$$ a_{i} = a_{j} + a_{k}, k \leq j < i $$

$ l(n) $ - наименьшая длина аддитивной цепочки для заданного $ n \in \mathbb{N} $.

Неравенство для метода множителей: $ \quad l(mn) \leq l(m) + l(n) $

Неравенство для m-арного метода $ [m = 2^k, n = \sum_{j=0}^{k} d_{j}m^{t-j}] $: $ \quad l(n) = m - 2 + (k+1)t $

Неравенство для бинарного алгоритма "SX": $ \quad l(n) = \lambda(n) + \nu(n) - 1 $

### Свойства аддитивных цепочек
* Полагается, что все аддитивные цепочки - возрастающие:
$ 1 = a_0 < a_1 < ... < a_r = n $
* Если два числа из последовательности $\{a_i\}$ совпадают, то одно из них может быть опущено
* Пара $ (j, k), 0 \leq k < j < i $ называется шагом
* Если существует более одной пары $ (j, k) $, полагается, что $ j $ - наибольшее из возможных 

### Виды шагов 
* удвоение: $ \quad j = k = i - 1 $
* звездный шаг: $ \quad j = i - 1 $
* малый шаг: $ \quad \lambda(a_{i}) = \lambda(a_{i-1}) $

### Свойства шагов
* Первый шаг - всегда удвоение
* Удвоение - звездный шаг, но никогда не малый шаг
* За удвоением всегда следует звездный шаг
* Если $ i-й $ шаг не малый, то $ (i + 1)-й $ - либо звездный, либо малый либо и тот, и другой
* Если $ (i + 1)-й $ шаг ни звездный, ни малый, то $ i-й $ шаг должен быть малым

### Теорема
Если аддитивная цепочка включает $ d $ удвоений и $ f = r - d $ неудвоений, то 
$$ n \leq 2^{d-1}F_{f+3}, $$
где $ F_j $ - число Фибоначчи под номером $ j $.
### Следствие из теоремы
Если аддитивная цепочка содержит $ f $ удвоений и $ S $ малых шагов, то
$$ S \leq f \leq \frac{S}{1-log_2(\varphi)}, $$ 
где $ \varphi = \frac{\sqrt{5} + 1}{2} $ - золотое сечение.

### Алгоритм Брауэра
Для заданного числа $ n \in \mathbb{N} $ можно построить цепочку Брауэра с помощью рекуррентной формулы
$$ B_k(n) = 
\begin{cases} 
1, 2, 3, ..., 2^{k} - 1,\  если \  n < 2^{k}\\
B_k(q), 2q, ..., 2^kq, n,\  если \  n \geq 2^{k} и \  q = \lfloor{\frac{n}{2^k}} \rfloor
\end{cases} $$

Данная цепочка будет иметь длину
$$ l_B(n) = j(k + 1) + 2^k - 2 $$
при условии, что $ jk \leq \lambda(n) < (j + 1)k $

Длина цепочки будет минимизирована для больших $ n $, если 
$$ k = \lambda(\lambda(n)) - 2\lambda(\lambda(\lambda(n))) $$

#### Уточнения
* Задается фикированный параметр $ k $. Выполняются вычисления вспомогательных чисел:
$$ 
d = 2^k, q_1 = [\frac{n}{d}], r_1 = n \ mod \ d \Rightarrow n = q_{1}d + r_{1} \ (0 \leq r_{1} < d) \\
q_2 = [\frac{q_1}{d}], r_2 = q_1 \ mod \ d \Rightarrow q_1 = q_{2}d + r_{2}
$$
* Данная процедура выполняется до тех пор, пока $ q_s $ не станет меньше $ d $. Следовательно $ q_{s-1} = q_{s}d + r_{s} $
* Таким образом,
$$ n = 2^kq_1 + r_1 = ... = 2^k(2^k(...(2^kq_s + r_s)...) + r_2) + r_1 $$
$$ B_k(n): 1, 2, 3, ..., 2^k - 1, 2q_s, 4q_s, ..., 2^kq_s + r_s, ..., n $$





## Постановка задачи

Реализовать точные и приближённые алгоритмы нахождения минимальных аддитивных цепочек с использованием системы компьютерной математики SageMath, провести анализ алгоритмов. Полученные результаты содержательно проинтерпретировать.

## Выполнение работы
#### 1. 
Вручную (т.е. **не реализовывая алгоритм** на Sage) построить последовательность вычислений бинарным методом и методом множителей для $ x^n $ для 2-3 значений 𝑛 (значение $ n \geq 30 $ выбираются студентом самостоятельно). Сравнить количество операций для каждого метода, сделать выводы.

**n = 40**

`Бинарный метод`

|   | N  | Y    | Z    |
|---|----|------|------|
| 0 | 40 | 1    | x    |
| 1 | 20 | 1    | x^2  |
| 2 | 10 | 1    | x^4  |
| 3 | 5  | 1    | x^8  |
| 4 | 2  | x^8  | x^16 |
| 5 | 1  | x^8  | x^32 |
| 6 | 0  | x^40 | x^32 |

`Метод множителей`


$ 40 = 2 \cdot 20 $,
$$ x^{40} = (x^{20})^{2} $$
$ 20 = 2 \cdot 10 $,
$$ x^{20} = (x^{10})^{2} $$ 
$ 10 = 2 \cdot 5 $,
$$ x^{10} = (x^{5})^{2} $$
5 - простое число,
$$ x^{5} = x \cdot x^{4} $$
$ 4 = 2 \cdot 2 $,
$$ x^{4} = (x^{2})^{2} $$

$$ x^{40} = (((x^{2})^{2})^{2})^{5} $$

Аддитивная цепочка: $ 1, 2, 4, 8, 16, 32, 40 $

**n = 41**

`Бинарный метод`

|   | N  | Y    | Z    |
|---|----|------|------|
| 0 | 41 | 1    | x    |
| 1 | 20 | x    | x^2  |
| 2 | 10 | x    | x^4  |
| 3 | 5  | x    | x^8  |
| 4 | 2  | x^9  | x^16 |
| 5 | 1  | x^9  | x^32 |
| 6 | 0  | x^41 | x^32 |

`Метод множителей`


41 - простое число,
$$ x^{41} = x \cdot x^{40} $$
$ 40 = 2 \cdot 20 $,
$$ x^{40} = (x^{20})^{2} $$
$ 20 = 2 \cdot 10 $,
$$ x^{20} = (x^{10})^{2} $$ 
$ 10 = 2 \cdot 5 $,
$$ x^{10} = (x^{5})^{2} $$
5 - простое число,
$$ x^{5} = x \cdot x^{4} $$
$ 4 = 2 \cdot 2 $,
$$ x^{4} = (x^{2})^{2} $$

$$ x^{41} = x  ( ( (x^{2})^{2})^{2})^{5} $$

Аддитивная цепочка: $ 1, 2, 4, 8, 16, 32, 40, 41 $

**Вывод**

Проанализировав данные примеры, можно сделать вывод, что если в числе количество единиц в двоичной записи превышает количество нулей, то бинарный метод работает в среднем медленнее, чем метод множителей. Если же количество нулей в двоичной записи числа преобладает, то бинарный метод работает за меньшее число операций, чем метод множителей.

#### 2.
Реализовать алгоритм Брауэра для вычисления приближённых аддитивных цепочек для различных чисел при варьировании параметра 𝑘, сопоставить длины полученных аддитивных цепочек с минимальной аддитивной цепочкой для заданного числа. Сделать выводы.

In [8]:
def brauer():
    n=int(input())
    k=int(input())
    d=2^k
    r=[]
    leftovers=[]
    subsqence=[]
    buff=0
    coefficients=0
    while(n>d):
        r.append(n%d)
        n=n//d
        leftovers.append(n)
    
    for i in range(1, d):
        subsqence.append(i)
        
        
    for i in range(1, len(r)+1):
        coefficients=2
        buff=leftovers[-i]
        while coefficients <= d:
            subsqence.append(leftovers[-i]*coefficients)
            coefficients = coefficients*2
        subsqence.append(buff*coefficients/2+r[-i])
        
    print(sorted(subsqence))
        
brauer()

31415
3
[1, 2, 3, 4, 5, 6, 7, 14, 28, 56, 61, 122, 244, 488, 490, 980, 1960, 3920, 3926, 7852, 15704, 31408, 31415]


Из работы программы можно сделать вывод, что, чтобы алгоритм Брауэра строил цепочки относительно небольшой длины, k необходимо задавать сильно меньше n, а для небольших n лучше брать k = 1 или k = 2.

#### 3.
Реализовать алгоритм дробления вектора индексов для нахождения минимальной звёздной цепочки для заданного числа. Протестировать алгоритм минимум для 5 значений 𝑛 > 1000. Указать, сколько времени потребовалось на поиск цепочки и какая цепочка получилась. Сравнить с предыдущими методами, сделать выводы.

In [6]:
from math import floor, log
from datetime import datetime


def get_chain(vec):  
    line = []
    line.append(1)
    for i in range(len(vec)):
        line.append(line[i] + line[vec[i] - 1])
    return line


def dect(vec): 
    for i in vec:
        if i > 1:
            return False
    return True

def dec(vec):
    _len = len(vec)
    for i in range(_len):
        if vec[_len - i - 1] > 1:
            vec[_len - i - 1] -= 1
            for j in range(_len - i, _len):
                vec[j] = j + 1
            break





def vector_index(n):
    
    BINAR = format(n, 'b')
    LAMBDA = len(BINAR) - 1
    NU = BINAR.count('1')
    MAX_l = LAMBDA + NU - 1
    MIN_l = LAMBDA + floor(log(n, 2))
    tries = [1, 1]

    for m in range(MAX_l, MIN_l):
        q = int(m / 2)

        vector_part = [0] * q
        for i in range(q):
            vector_part[i] = 1 + i

        vector_fix = vector_part

        while True:  

            change_part = [0] * (m - q)
            for i in range(m - q):
                change_part[i] = q + 1 + i

            change = change_part

            bounds = [0] * 2

            a = []
            a.append(1)
            for i in range(len(vector_fix)):
                a.append(a[i] + a[vector_fix[i] - 1])
            a_1 = a[len(a) - 1]

            bounds[0] = a_1 + m - q
            bounds[1] = a_1 * 2 ** (m - q)

            if n < bounds[0] or n > bounds[1]:
                if dect(vector_fix) and len(vector_fix) > 1:
                    break
                dec(vector_fix)
                continue

            while True:  

                vector_full = []
                vector_full.append(1)
                for i in range(len(vector_fix + change)):
                    vector_full.append(vector_full[i] + vector_full[(vector_fix + change)[i] - 1])
                vector_full = vector_full[len(vector_full) - 1]

                if vector_full == n:
                    return vector_fix + change

                if dect(change):
                    break
                dec(change)
            if dect(vector_fix):
                break
            dec(vector_fix)


n = int(input())
time_1 = datetime.now()
vova = vector_index(n)

print("chain : ", get_chain(vova))
time_2 = datetime.now()
print(time_2-time_1)

1974
chain :  [1, 2, 4, 8, 16, 32, 64, 128, 256, 512, 1024, 1536, 1540, 1556, 1588, 1590, 1718, 1974]
0:00:02.853319


|   n   |                                                a                                                  | len   | time  |
|-------|---------------------------------------------------------------------------------------------------|-------|-------|
| 3571  | 1, 2, 4, 8, 16, 32, 64, 128, 256, 512, 1024, 2048, 3072, 3074, 3090, 3122, 3186, 3314, 3570, 3571 | 20    | 0,12  |
| 1536  | 1, 2, 4, 8, 16, 32, 64, 128, 256, 512, 1024, 1536                                                 |  18   | 0,02  |
| 2002  | 1, 2, 4, 8, 16, 32, 64, 128, 192, 384, 768, 1536, 1920, 1936, 1968, 2000, 2002                    |  17   | 0,02  |
| 1974  | 1, 2, 4, 8, 16, 32, 64, 128, 256, 512, 1024, 1536, 1540, 1556, 1588, 1590, 1718, 1974             |  18   | 0,02  |


4)Проверим гипотезу Штольца-Брауэра

Результаты проверки гипотезы Шольца-Брауэра для $$1\leq n \leq 12$$  представлены в таблице ниже:

| n  | $$l(2^{n - 1}) $$ | l(n) + n - 1    | Гипотеза Шольца-Брауэра  |
|----|-------------------|-----------------|--------------------------|
| 1  | 0                 | 0               | x^2                      |
| 2  | 2                 | 2               | x^4                      |
| 3  | 4                 | 4               | x^8                      |
| 4  | 5                 | 5               | x^16                     |
| 5  | 7                 | 7               | x^32                     |
| 6  | 8                 | 8               | x^32                     |
| 7  | 10                | 10              | x^2                      |
| 8  | 10                | 10              | x^4                      |
| 9  | 12                | 12              | x^8                      |
| 10 | 13                | 13              | x^16                     |
| 11 | 15                | 15              | x^32                     |
| 12 | 15                | 15              | x^32                     |


Отсюда можно сделать вывод, что гипотеза Шольца-Брауэра выполняется для всех N от 1 до 12

# Вывод

В ходе выполнения практической работы были реализованы алгоритмы для создания аддитивных цепочек алгоритмом Брауэра и алгоритмом дробления вектора индексов. Исследования показали, что алгоритм Брауэра лучше использовать с достаточно малым параметром k относительно n, а при малых n лучше брать k=1. Алгоритм дробления вектора индексов строит минимальную звёздную цепочку, но затрачивает на это достаточно большое количество времени. Гипотеза Шольца-Брауэра успешно подтверждена на алгоритме дробления вектора индексов.